src/defrag.c

```diff
@@ -410,87 +392,79 @@ long scanLaterList(robj *ob, unsigned long *cursor, long long endtime, long long
 
 typedef struct {
     zset *zs;
-    long defragged;
 } scanLaterZsetData;
 
 void scanLaterZsetCallback(void *privdata, const dictEntry *_de) {
     dictEntry *de = (dictEntry*)_de;
     scanLaterZsetData *data = privdata;
-    data->defragged += activeDefragZsetEntry(data->zs, de);
+    activeDefragZsetEntry(data->zs, de);
     server.stat_active_defrag_scanned++;
 }
 
-long scanLaterZset(robj *ob, unsigned long *cursor) {
+void scanLaterZset(robj *ob, unsigned long *cursor) {
     if (ob->type != OBJ_ZSET || ob->encoding != OBJ_ENCODING_SKIPLIST)
-        return 0;
+        return;
     zset *zs = (zset*)ob->ptr;
     dict *d = zs->dict;
-    scanLaterZsetData data = {zs, 0};
+    scanLaterZsetData data = {zs};
     *cursor = dictScanDefrag(d, *cursor, scanLaterZsetCallback, activeDefragAlloc, &data);
-    return data.defragged;
 }
 
 typedef struct {
     dict *dict;
-    long defragged;
 } scanLaterDictData;
 
 void scanLaterSetCallback(void *privdata, const dictEntry *_de) {
     dictEntry *de = (dictEntry*)_de;
     scanLaterDictData *data = privdata;
     sds sdsele = dictGetKey(de), newsds;
     if ((newsds = activeDefragSds(sdsele)))
-        data->defragged++, dictSetKey(data->dict, de, newsds);
+        dictSetKey(data->dict, de, newsds);
     server.stat_active_defrag_scanned++;
 }
 
-long scanLaterSet(robj *ob, unsigned long *cursor) {
+void scanLaterSet(robj *ob, unsigned long *cursor) {
     if (ob->type != OBJ_SET || ob->encoding != OBJ_ENCODING_HT)
-        return 0;
+        return;
     dict *d = ob->ptr;
-    scanLaterDictData data = {d, 0};
+    scanLaterDictData data = {d};
     *cursor = dictScanDefrag(d, *cursor, scanLaterSetCallback, activeDefragAlloc, &data);
```

Comment:
Now we could actually just use the `dict *` as privdata and cut a few more lines. Similar for other scan data. Should I ...?

> WOW.. ~500 lines, that came out much bigger than i expected.

Yes, you thought it's "some 30 lines" :laughing: I think the code got much cleaner without passing around the counter and I don't think we should avoid making simplifications just for the git blame.

Reply:
yes, i only counted the lines doing the increments, and not the ones passing the pointer, and function declarations.
apparently when looking at a commit, git GH line stats still show the whole RP.

not sure we have to proceed cleaning the struct. maybe we'll need them someday soon. your call.

either way since we have several different topics in this PR, each with quite a lot of changed lines, i'd like to separate that change from others, so anyone reading the diff can focus on the important things (like the other changes we did in defrag.c and dict.c)

Reply:
we can probably do some git squashing and rebasing effort when that PR is ready.
the other alternative is to start a new branch, and copy changes from the old one in stages, but that's gonna take some work to find them.
note that i don't think we have to make sure each commit stands for itself and compiles.
i want it mainly for the purpose of code review (now and in the future), so it's easier to focus on important changes by topic without running into a huge refactoring diff.

taking it into a separate PR is also a possibility. but maybe there would still be other things in this PR that are worth splitting?

FYI, some examples where we did such things (on bigger projects): #9320, #9780, #7807

Reply:
OK, I squashed them down to 6 commits. I think this is ready, unless you have other comments.

I guess you don't want to merge it until there is use for it, so I'll start working on some kind dict improvement based on this. :-)